Reinitialize oidc provider on refresh token rotation

slaskawi opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

<!--
Add one of the following kinds:
kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This Pull Request ensures the `newOIDCAuthProvider` creates a new client whenever the refresh token changes in the `cfg`. The implementation is inspired by the @liggitt 's comment here https://github.com/kubernetes/kubernetes/pull/78509#discussion_r310296242

This Pull Request is also related to https://github.com/kubernetes/kubernetes/pull/78509

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #78494

#### Special notes for your reviewer:

Please have a look at https://github.com/kubernetes/kubernetes/issues/78494 prior to reviewing this Pull Request. 

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

slaskawi committed:
Reinitialize oidc provider on refresh token rotation

liggitt commented:
/cc @enj @tallclair 

if @enj or @tallclair could take a look first that would be helpful

tallclair reviewed (commented):
https://github.com/kubernetes/kubernetes/blob/718e1fb23ec8cf5ad70c7f7447b6f683c4e43d39/staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/oidc.go#L108-L111

What are the implications of this for rotating the client? Does the client mutex need to be shared across cache invalidations?

tallclair commented:
/ok-to-test

slaskawi commented:
> What are the implications of this for rotating the client? Does the client mutex need to be shared across cache invalidations?

@tallclair Regarding to the mutex, we should be fine. The `deleteClient` uses the same one as the `setClient`. Even though I see the explanation in the comment, returning an old client when calling `setClient` seems a bit weird. 

@deads2k this code is used only in `newOIDCAuthProvider` and it not exported, do we need this piece? `git gui blame` tells me you added it. Side note - we also have a test verifying if the old client is indeed returned, so this piece is definitely intentional.

slaskawi commented:
@tallclair Comments addressed. Ready for the next round of review.

enj commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/oidc.go:
```diff
@@ -121,7 +130,12 @@ func newOIDCAuthProvider(clusterAddress string, cfg map[string]string, persister
 
 	// Check cache for existing provider.
 	if provider, ok := cache.getClient(clusterAddress, issuer, clientID); ok {
-		return provider, nil
+		// Return the provider only if the refresh token hasn't changed. Otherwise,
+		// remove the stale client from the cache and let it re-initialize further
+		if cfg[cfgRefreshToken] == provider.cfg[cfgRefreshToken] {
+			return provider, nil
+		}
+		cache.deleteClient(clusterAddress, issuer, clientID)
```
This is not thread safe, i.e. it suffers from TOCTOU.  More than one go routine could go into the `if` statement, and the second could unnecessarily delete the client.

slaskawi replied:
Thanks for pointing this out @enj! 

Probably the easiest way to fix it is to use `cache.mu.Lock` around both the `cache.getClient` and `cache.deleteClient` calls. 

But maybe it's better to use an approach suggested in https://github.com/kubernetes/kubernetes/pull/78509 ? Reuse the client from the cache but update its configuration? I think it's a slightly better solution than creating a long critical section here. WDYT?

slaskawi replied:
A side note from looking into this a bit more - the TOCTOU bug always there. Just imagine two Go Routines going parallel through the `if provider, ok := cache.getClient(clusterAddress, issuer, clientID)` check.

slaskawi replied:
I introduced a new mutex to and created a critical section of the `newOIDCAuthProvider` content.

enj commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/oidc.go:
```diff
@@ -154,7 +168,7 @@ func newOIDCAuthProvider(clusterAddress string, cfg map[string]string, persister
 	provider := &oidcAuthProvider{
 		client:    hc,
 		now:       time.Now,
-		cfg:       cfg,
+		cfg:       deepCopyStringMap(cfg),
```
I need to think through what holding on to the original config means over the lifetime of a long running process.

slaskawi replied:
Any update on this? Are you OK with this approach?

tallclair commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/oidc_test.go:
```diff
@@ -22,6 +22,8 @@ import (
 	"testing"
 	"time"
+
+	"github.com/stretchr/testify/require"
```
I disagree, but I'm also a proponent of `testify` (I think it makes a lot of test errors more readable and easier to debug, and cuts a lot of boilerplate). Looks like it's currently used in 216 packages, although that's admittedly that's less than 10% of the total:
```sh
git grep -l "github.com/stretchr/testify" | grep ".*\.go" | xargs dirname | sort -u | wc -l
216
```

We should probably ask @kubernetes/sig-testing or the code organization group for an official stance on this, but for now I'd say the it's up to the author.

enj replied:
I generally share the sentiment regarding `testify`, but I favor consistency.  Very little of SIG auth's code uses so I would prefer an official stance on it over letting the author decide.

slaskawi replied:
No problem, I can just align this with the rest of SIG Auth code base.

slaskawi replied:
Removed testify.

tallclair commented:
>> What are the implications of this for rotating the client? Does the client mutex need to be shared across cache invalidations?
>
> Regarding to the mutex, we should be fine. The deleteClient uses the same one as the setClient. Even though I see the explanation in the comment, returning an old client when calling setClient seems a bit weird.

This comment was with regards to the mutex on the client, not on the cache. This one:
https://github.com/kubernetes/kubernetes/blob/2d9135e92b904218411b1a2914d039d2f910d279/staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/oidc.go#L184-L187

fedebongio commented:
/remove-sig api-machinery

slaskawi commented:
@tallclair @enj I addressed most of the comments. There are still some doubts in the comments but I think we are ready for the next round of the review.

tallclair reviewed (commented):
We need to think through the concurrency model here in a more holistic way. I think @enj has some ideas.

slaskawi commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/oidc.go:
```diff
@@ -109,6 +125,8 @@ func (c *clientCache) setClient(clusterAddress, issuer, clientID string, client
 
 func newOIDCAuthProvider(clusterAddress string, cfg map[string]string, persister restclient.AuthProviderConfigPersister) (restclient.AuthProvider, error) {
+	newOIDCAuthProviderMutex.Lock()
+	defer newOIDCAuthProviderMutex.Unlock()
```
Actually you do need both as locks in Golang are not reentrant (the same lock can not be acquired by the same Go routine twice). If I acquired a `cache.mu` lock here and then call `cache.deleteClient`, it will block forever (both will deadlock). So unless we have a reentrant lock implementation somewhere, we just need another lock.

And I do agree with you, if you have more than one lock, you need to be careful with locking order. A rule of thumb - acquire and release lock always in the same order.

So I'm open for suggestions how to improve this piece.

slaskawi commented on staging/src/k8s.io/client-go/plugin/pkg/client/auth/oidc/oidc.go:
```diff
@@ -121,7 +139,13 @@ func newOIDCAuthProvider(clusterAddress string, cfg map[string]string, persister
+			//cache.mu.Unlock()
+			return provider, nil
+		}
+		cache.deleteClient(clusterAddress, issuer, clientID)
```
This is actually what happens under the covers. The `deleteClient` grabs the write lock and then removes the client if it's still there (if someone already removed it - the `deleteClient` won't complain). 

So unless I misunderstood you, we're fine here, right?

enj commented:
The existing code (in `master`) has some difficult to reason about multi-leveled caching along with multiple mutexes.  I would like to think about how we can carefully refactor this code into a more understandable and maintainable state.  I do not believe we have enough time to get that into 1.24 since today is code freeze, so I am going to push this out to 1.25.

/milestone v1.25

cici37 commented:
/remove-sig api-machinery

slaskawi commented:
> The existing code (in master) has some difficult to reason about multi-leveled caching along with multiple mutexes. I would like to think about how we can carefully refactor this code into a more understandable and maintainable state. I do not believe we have enough time to get that into 1.24 since today is code freeze, so I am going to push this out to 1.25.

@enj I absolutely agree. Here are a few things I'd like to highlight before the refactoring:
- The `clientCache` is not exported, which means that the whole mutex logic is owned by the `oidc` module.
- Moreover, all `clientCache` methods are invoked only from `newOIDCAuthProvider`. This effectively means, the synchronization is only required at `newOIDCAuthProvider` level. We don't have to go deeper.

Having said that, I propose:
- Removing mutex from `clientCache` and adding a comment that the caller is responsible for locking strategy.
- (Optional) simplify `clientCache` and turn it to a `map[cacheKey]*oidcAuthProvider`. We don't need extra methods, such as `getClient` or `setClient`.
- Lock the whole `newOIDCAuthProvider` method with `newOIDCAuthProviderMutex`. This is the simplest approach as this function is quite long and constructing a `restclient.AuthProvider` involves multiple steps. 

Do you agree to this plan @enj and @tallclair ?

slaskawi commented:
@enj I updated (and rebased) the code. It is now much easier to reason about it in terms of concurrency. 

Here's the summary of changes:
* Removed `mutex` from `clientCache` and added a comment that the synchronization policy is delegated to the caller. 
* I decided to leave the `(get/set/delete)client` methods. They are heavily used in the testing code.
* The whole `newOIDCAuthProvider` is now guarded with a single `mutex`.

cici37 commented:
/remove-sig api-machinery

slaskawi commented:
Let me reassign this to @enj 

/assign @enj

enj commented:
@slaskawi sorry for the delay.

I have been thinking about this change and the associated code, especially in regards to the on-going [quality discussion](https://groups.google.com/a/kubernetes.io/g/dev/c/6F3h0Z1QzVg/m/43tdYs_lAAAJ).

While I think this change is at least partially a move in the right direction, I believe further refactors are required to actually make this code understandable and maintainable (and correct in all configurations).

Thus before making changes to this code, I would like to see the following integration test coverage:

- Build a mock OIDC test server with the JWKs, well-known, and token endpoints
- Run a test API server
- Configure it to use a the mock OIDC test server (process local)
- Configure the client(s) based on the specific test scenario
- Make calls to the CSR API _and_ the token review API (with and without correct audience) and assert the user info is correct

Some scenarios that come to mind (this is not meant to be a complete list, and I suspect some will fail on the existing code):

1. ID token not expired
2. ID token expired
3. Refresh token does not change
4. Refresh token does change
5. Other aspects of the config change (i.e. client secret, CA data, etc)
6. Auth config persister does not actually persist
7. Refresh flow does not return ID token
8. ID token expires early due to JWK rotation
9. Distributed claim fetching for group data
10. ...

The above will likely need multiple permutations, in particular I would like to see both kubeconfig files as input (maybe even the same file concurrently across different round trippers) and `rest.Config` as inputs (with every weird config a code integrator could come up with).

In an ideal world we would have e2e tests in a similar vain as the above against an OIDC provider like Dex or Keycloak or UAA, but we can leave that off for now.

The API server OIDC support and the client-go OIDC support are two halves of the same feature, and both sides have no real integration test coverage when used in conjunction with an API server.  Lets fix that so that we can make changes with confidence.

enj commented:
/hold

Just so that it is clear, I consider https://github.com/kubernetes/kubernetes/pull/108794#issuecomment-1126162050 as a blocking pre-req for this and any other changes to this code.

markyjackson-taulia commented:
Hi @slaskawi  Just checking in to see if this is still on track for k8s 1.25?

markyjackson-taulia commented:
Hello @enj, my name is Marky Jackson and I am one of the k8s 1.25 bug triage shadow assigned to track this body of work. Is this still on track for 1.25?

helayoty commented:
/milestone clear
